### PR TITLE
vg: migrate draw.Point+draw.Rectangle back to vg

### LINKED ISSUE #262: vg: migrate draw.Point and draw.Rectangle back to vg
this is needed by #256 

also: we need to adapt some of the wiki examples:
 https://github.com/gonum/plot/wiki/Creating-Custom-Plotters:-A-tutorial-on-creating-custom-Plotters

### PR DESCRIPTION
**Pull request by sbinet**
Fixes #262

### CODE REVIEW

**Comment by sbinet** [comment]
it's pure code movement.
*except* for some methods of `Point` that I had to export (`minus` -> `Minus`, `dot`, `scale` and `plus`) because they were used by `draw.Canvas` internally (to build `vg.Path`s)

**Comment by sbinet** [comment]
BTW, while we're at it.
perhaps it would be interesting to migrate `vg.Length`, `vg.Point` and `vg.Rectangle` to `x/image/math/fixed.Int52_12` and friends ?
 https://godoc.org/golang.org/x/image/math/fixed

this would probably ensure better compatibility and interoperability with other libraries.

what do you think?

**Comment by kortschak** [comment]
*If* there is a desire to change the dimension type, it should happen in another PR.

**Comment by sbinet** [comment]
agreed.

File: vg/geom.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	X, Y Length
+}
+
+// Dot returns the dot product of two points.
+func (p Point) Dot(q Point) Length {
+	return p.X*q.X + p.Y*q.Y
+}
+
+// Plus returns the component-wise sum of two points.
+func (p Point) Plus(q Point) Point {
```
Comment (kortschak) [inline vg/geom.go]: Since we're exposing these for the first time, how about we choose more commonly used names.

`Add`

File: vg/geom.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	return Point{p.X + q.X, p.Y + q.Y}
+}
+
+// Minus returns the component-wise difference of two points.
+func (p Point) Minus(q Point) Point {
```
Comment (kortschak) [inline vg/geom.go]: `Sub`

**Comment by kortschak** [comment]
LGTM after consideration of names suggestion, waiting for @eaburns.

**Comment by kortschak** [comment]
@jonlawlor I take it these tests aren't being run on travis because of my paranoia?

File: vg/geom.go (hunk 1 above)
Comment (sbinet) [inline vg/geom.go]: done

File: vg/geom.go (hunk 2 above)
Comment (sbinet) [inline vg/geom.go]: done

**Comment by sbinet** [comment]
FTR, I just noticed it as well and went to the travis dashboard for gonum/plot and re-enabled builds on push requests.
Hope that's ok.

File: vg/draw/canvas.go [hunk 3]
```diff
@@ -549,30 +549,30 @@ func clipPoly(in func(Point, Point) bool, clip, norm Point, pts []Point) (clippe
...
 	return p.Y >= clip.Y-slop
 }
 
 // isect returns the intersection of a line p0→p1 with the
 // clipping line specified by the clip point and normal.
-func isect(p0, p1, clip, norm Point) Point {
+func isect(p0, p1, clip, norm vg.Point) vg.Point {
 	// t = (norm · (p0 - clip)) / (norm · (p0 - p1))
-	t := p0.minus(clip).dot(norm) / p0.minus(p1).dot(norm)
+	t := p0.Sub(clip).Dot(norm) / p0.Sub(p1).Dot(norm)
```
Comment (eaburns) [inline vg/draw/canvas.go]: I prefer naming these nominally, not verbally. They aren't performing an action, they are returning a thing. `Sub` is not subtracting from `p0`. The value of `p0` remains fixed. It's returning the difference of `p0` and `clip`. We call that difference "`p0` minus `clip`", which in code looks like this: `p0.Minus(clip)`.

Comment (eaburns) [inline vg/draw/canvas.go]: On the other hand, I don't know what to do about Scale.

Comment (eaburns) [inline vg/draw/canvas.go]: I see that @kortschak told you to change these. I don't want to argue over these names, so just leave them.

**Comment by eaburns** [comment]
LGTM.

File: vg/draw/canvas.go (hunk 3 above)
Comment (sbinet) [inline vg/draw/canvas.go]: one could create the following free functions:
```go
func SubPoints(p1,p2 Point) Point {...}
func AddPoints(p1,p2 Point) Point {...}
func DotPoints(p1,p2 Point) Length {...}
func ScalePoint(s Length, p Point) Point {...}
```

Comment (eaburns) [inline vg/draw/canvas.go]: That makes their use a lot more verbose. These names are fine. They are mostly for internal use anyway.

**Comment by jonlawlor** [comment]
@kortschak I don't know why the commit doesn't seem to have built on travis.  There are other commits and pull requests that are either running right now or ran in the last 20 minutes.  Also checking the settings for the repo and all pushes and pull requests are set to build.

Also I just saw sbinet's commits & tests go through a minute ago, but github is listing them as 5 hours old, maybe there was a big delay between commit and push?

File: vg/draw/canvas.go (hunk 3 above)
Comment (kortschak) [inline vg/draw/canvas.go]: The suggestion was based on the naming of time.Time.Sub which returns a time.Time.

**Comment by sbinet** [comment]
ok. 2 LGTMs w/ Sub, Add and friends.

before merging this into `master`, should we send a quick note (at least on `gonum-dev` and perhaps even on `golang-nuts`) about the breakage and the mechanics for the migration ?
or wait for #266 and its associated breakage before merging everything into `master`.

**Comment by kortschak** [comment]
Send a post notifying of impending breakage, but do all the merges together I think.

**Pull request merged**